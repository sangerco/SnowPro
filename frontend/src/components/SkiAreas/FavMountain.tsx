import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { fetchFavMountainData, deleteFavMountain } from '../../redux/actions/favMountainActions';
import { Container, Card, Button, Icon } from 'semantic-ui-react';
import { RootState } from '../../redux/store';

interface FavMountainData {
    id: string;
    userId: string;
    skiAreaSlug: string;
    skiAreaName: string;
}

interface FavMountainProps {
    userId: string;
    favMountain: FavMountainData[] | null;
    error: string | null;
    fetchFavMountainData: (userId: string) => void
    deleteFavMountain: (id: string) => void
};

const FavMountain: React.FC<FavMountainProps> = ({ userId, favMountain, error, fetchFavMountainData, deleteFavMountain }) => {
    useEffect(() => {
        fetchFavMountainData(userId);
    }, [userId, fetchFavMountainData]);

    if(error) {
        return <p>{`Error fetching Favorite Mountain Data! Error: ${error}`}</p>
    };

    if(favMountain) {
        return (
            <Container fluid>
                {favMountain.length > 0 ? 
                    favMountain.map(fav => (
                        <Card key={fav.skiAreaSlug}>
                            <Card.Header>{fav.skiAreaName}</Card.Header>
                            <Card.Content extra>
                                <Button onClick={() => deleteFavMountain(fav.id)}>
                                    <Icon name='cancel' />
                                </Button>
                            </Card.Content>
                        </Card>
                    ))
                    :
                    <p>No Favorite Mountains Yet!</p>}
            </Container>
        )
    }

    return null;
};

const mapStateToProps = (state: RootState) => ({
    favMountain: state.favMountain.data,
    error: state.favMountain.error
});

const mapDispatchToProps = {
    fetchFavMountainData,
    deleteFavMountain
};

export default connect(mapStateToProps, mapDispatchToProps)(FavMountain);