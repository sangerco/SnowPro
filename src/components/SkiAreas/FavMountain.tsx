import React, { useEffect, useState } from "react";
import { Button, Card, Header, Icon, Modal, Segment } from "semantic-ui-react";
import { Link } from "react-router-dom";
import { fetchOneSkiArea } from "../../redux/slices/skiAreaSlice";
import {
  removeFavMountain,
  FavMountainData,
} from "../../redux/slices/favMountainSlice";
import { RootState, AppDispatch } from "../../redux/store";
import { useSelector, useDispatch } from "react-redux";

interface FavMountainProps {
  slug: string;
  username: string;
}

const FavMountain: React.FC<FavMountainProps> = ({ slug, username }) => {
  const dispatch = useDispatch<AppDispatch>();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const loggedInUsername = auth.data?.username;

  useEffect(() => {
    dispatch(fetchOneSkiArea(slug));
  }, [dispatch, slug]);

  const skiAreas = useSelector((state: RootState) => state.skiAreas);
  const skiArea = skiAreas.skiArea;
  const [showDeleteModal, setShowDeleteModal] = useState(false);

  const handleShowDeleteModal = () => {
    setShowDeleteModal(true);
  };

  const favMountainData: FavMountainData = {
    username: username ?? "",
    userId: userId ?? "",
    skiAreaSlug: slug,
  };

  const handleRemoveFavMountain = () => {
    if (skiArea && username) {
      dispatch(removeFavMountain(favMountainData));
    }
  };

  if (skiArea) {
    return (
      <>
        <Segment raised style={{ width: "250px", margin: "10px" }}>
          <Header as="h4" textAlign="left" data-testid="fm-header">
            <Link to={`/ski-areas/${slug}`}>{skiArea.name}</Link>
          </Header>
          {username === loggedInUsername ? (
            <Button
              data-testid="delete-button"
              floated="right"
              color="red"
              size="mini"
              onClick={handleShowDeleteModal}>
              <Icon name="times" />
            </Button>
          ) : null}
        </Segment>
        <Modal
          open={showDeleteModal}
          onClose={() => setShowDeleteModal(false)}
          data-testid="delete-modal">
          <Modal.Content>
            Remove this Ski Area from your favorites list?
          </Modal.Content>
          <Modal.Actions>
            <Button negative onClick={handleRemoveFavMountain}>
              Yes
            </Button>
            <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
          </Modal.Actions>
        </Modal>
      </>
    );
  }

  return null;
};

export default FavMountain;
