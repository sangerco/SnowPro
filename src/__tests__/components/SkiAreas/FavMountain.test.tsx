/* eslint-disable testing-library/no-node-access */
import React from "react";
import FavMountain from "../../../components/SkiAreas/FavMountain";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test FavMountain", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
        username: "testuser",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <FavMountain slug={"ski-area-1"} username={"testuser"} />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <FavMountain slug={"ski-area-1"} username={"testuser"} />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should show a favorite mountain", async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <FavMountain slug={"ski-area-1"} username={"testuser"} />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const header = screen.getByTestId("fm-header");
      expect(header).toBeInTheDocument();
    });
  });
  it("should show the delete modal", async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <FavMountain slug={"ski-area-1"} username={"testuser"} />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const button = screen.getByTestId("delete-button");
      // eslint-disable-next-line testing-library/no-wait-for-side-effects
      fireEvent.click(button);
    });

    await waitFor(() => {
      const modal = screen.getByTestId("delete-modal");
      expect(modal).toBeInTheDocument();
    });
  });
});
