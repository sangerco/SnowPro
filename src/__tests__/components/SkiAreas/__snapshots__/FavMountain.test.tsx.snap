// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test FavMountain should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui raised segment"
    style="width: 250px; margin: 10px;"
  >
    <h4
      class="ui left aligned header"
      data-testid="fm-header"
    >
      <a
        href="/ski-areas/ski-area-1"
      >
        Ski Area 1
      </a>
    </h4>
    <button
      class="ui red mini right floated button"
      data-testid="delete-button"
    >
      <i
        aria-hidden="true"
        class="times icon"
      />
    </button>
  </div>
</DocumentFragment>
`;
